packages/services-context/src/index.ts

```diff
@@ -88,12 +90,32 @@ export class Services implements ServicesInterface {
 
   private async initializeWalletServices(): Promise<WalletServices> {
     const { walletId, fullViewingKey, idbVersion: dbVersion } = await this.config;
-    const params = await this.querier.app.appParams();
-    if (!params.sctParams?.epochDuration) throw new Error('Epoch duration unknown');
-    const {
-      chainId,
-      sctParams: { epochDuration },
-    } = params;
+
+    // read params from storage
+    const storedParams = await localExtStorage
+      .get('params')
+      .then(j => j && AppParameters.fromJson(j));
+    // try to fetch params from network
+    const queriedParams = await this.querier.app.appParams().catch((e: unknown) => void e);
+
+    // prefer fetched params
+    const chainId = queriedParams?.chainId ?? storedParams?.chainId;
+    const epochDuration =
+      queriedParams?.sctParams?.epochDuration ?? storedParams?.sctParams?.epochDuration;
```

Comment:
Can we move the asserts closer to where the variables are set? Something like:

```ts
    const chainId = queriedParams?.chainId ?? storedParams?.chainId;
    if (!chainId) throw new Error('Cannot initialize without chainId');
    // fail if chainId is mismatched
    if (storedParams?.chainId && storedParams.chainId !== chainId)
      throw new Error(
        'Stored chainId does not match the remote chainId. Your local state may\
        be invalid, and you cannot use this remote endpoint without clearing\
        your cached chain.',
      );

    const epochDuration =
      queriedParams?.sctParams?.epochDuration ?? storedParams?.sctParams?.epochDuration;
    if (!epochDuration) throw new Error('Cannot initialize without epoch duration');

    // store chainId if it was missing
    if (!storedParams?.chainId && queriedParams?.chainId) {
      await localExtStorage.set('params', queriedParams.toJson() as Jsonified<AppParameters>);
    }
```

Reply:
idb can start up without epochDuration, so they are deliberately separated. i'll pull out the params request into a method to simplify it conceptually